### PR TITLE
Purchases: Perform redirect when user can't access Cancel Purchase page

### PR DESCRIPTION
**Pull request by gziolo**
Part of #275. This should fix also issue from #1025 for Cancel Purchase page only.

Whenever user tries to access Cancel Purchase page for the purchase that has been removed from the account we should redirect to the purchases list.
Whenever user tries to access Cancel Purchase page for the purchase that can't be cancelled we should redirect to the Manage Purchase page.

### Testing.
1. Go to http://calypso.localhost:3000/purchases.
2. Select purchase that can't be cancelled.
3. When `Manage Purchase` page is loaded append `/cancel` to the url.
4. You should get redirected back to the Manage Purchase page. 
5. Repeat (3), but this time also change purchase id in the url to some random value.
6. You should get redirected to the page with purchases list.

/cc @stephanethomas 

### Review
* [x] QA review
* [x] Code review

### CODE REVIEW

**Comment by fabianapsimoes** [comment]
Works as described, but before being redirected I do see the skeleton for Cancel Purchase for quite a while:

domain

Is there a way we can skip that, and redirect directly?

**Comment by gziolo** [comment]
> before being redirected I do see the skeleton for Cancel Purchase for quite a while
Is there a way we can skip that, and redirect directly?

I don't think we can skip that, @scruffian, @drewblaisdell or @stephanethomas any thoughts on this

**Comment by scruffian** [comment]
I don't think there's a way round it either.

File: client/me/purchases/cancel-purchase/index.jsx [hunk 1]
```diff
@@ -18,41 +18,71 @@ import Main from 'components/main';
...
+	redirect( props ) {
+		if ( this.state.isRedirecting ) {
+			return;
+		}
+
+		const purchase = getPurchase( props ),
+			selectedSite = getSelectedSite( props );
+		let redirectPath = paths.list();
+
+		if ( selectedSite && purchase ) {
```
Comment (drewblaisdell) [inline client/me/purchases/cancel-purchase/index.jsx]: Super minor, but we can combine these conditions:
`selectedSite && purchase && ! isCancelable( purchase )`

Comment (gziolo) [inline client/me/purchases/cancel-purchase/index.jsx]: :+1:

File: client/me/purchases/cancel-purchase/index.jsx [hunk 2]
```diff
@@ -18,41 +18,69 @@ import Main from 'components/main';
...
+		const purchase = getPurchase( props ),
+			selectedSite = getSelectedSite( props );
+
+		return ( selectedSite && purchase && isCancelable( purchase ) );
 	},
 
-	componentDidUpdate() {
-		this.ensurePageCanLoad();
+	redirect( props ) {
+		if ( this.state.isRedirecting ) {
```
Comment (drewblaisdell) [inline client/me/purchases/cancel-purchase/index.jsx]: Instead of adding state to an otherwise stateless component, can we just check if the data is invalid here?

```js
if ( ! this.isDataValid() ) {
  return null;
}
```

**Comment by drewblaisdell** [comment]
QA and code are good. To avoid adding state to a stateless component, I added a commit to check `isDataValid` instead of flipping a boolean in state to prevent an invalid render or double redirect. If you're cool with that change, this LGTM.

File: client/me/purchases/cancel-purchase/index.jsx [hunk 3]
```diff
@@ -18,41 +18,57 @@ import Main from 'components/main';
...
+		} else {
+			this.redirect( this.props );
+		}
 	},
 
 	componentWillReceiveProps( nextProps ) {
-		recordPageView( 'cancel_purchase', this.props, nextProps );
+		if ( this.isDataValid( nextProps ) ) {
+			recordPageView( 'cancel_purchase', this.props, nextProps );
+		} else if ( this.isDataValid( this.props ) && ! this.isDataValid( nextProps ) ) {
```
Comment (gziolo) [inline client/me/purchases/cancel-purchase/index.jsx]: We can simplify this line to:
```js
} else if ( this.isDataValid( this.props ) ) {
```
but let's put it in another PR :)

Comment (drewblaisdell) [inline client/me/purchases/cancel-purchase/index.jsx]: Nice catch.
